charms/istio-pilot/src/charm.py

```diff
@@ -319,8 +362,29 @@ def upgrade_charm(self, _):
 
         Supports upgrade of exactly one minor version at a time.
         """
-        istioctl = Istioctl(ISTIOCTL_PATH, self.model.name, ISTIOCTL_DEPOYMENT_PROFILE)
-        self._log_and_set_status(MaintenanceStatus("Upgrading Istio"))
+        upgrade_message = "Upgrading Istio control plane."
+        # Verify the CNI configurations are set, otherwise we cannot continue with the
+        # Control Plane installation. Deferring the event is the only way we can prevent
+        # an attempt to install the Control Plane w/o these required values.
+        if not self._check_cni_configurations():
+            self.log.warning(
+                "This version of istio-pilot requires the cni-bin-dir and cni-conf-dir "
+                "configurations to be set in order to enable the Istio CNI plugin. "
+                "The installation with the Istio CNI plugin will be skipped. "
+                "If you want to enable this feature, please provide the required configuration."
+            )
+            upgrade_message = (
+                "Upgrading Istio control plan without Istio CNI plugin. "
+                "See `juju debug-log` for details."
+            )
```

Comment:
If this message is really needed, we should make it common with the message displayed in install.  They can come from the same constant or helper function.

I'm not sure this message adds much, and if we also add a call to `upgrade_charm` in reconcile ([here](https://github.com/canonical/istio-operators/pull/365/files#diff-d3a4e684bd1a3a1822df5ab7605cfce00c97cd7b52f44ed770526c4f4f80eacdR293)), the logs are going to have this warning on every config-changed, relation-joined, relation-changed, etc.    That feels like a lot

Reply:
Yeah, I kind of also wanted to abstract it to a common helper or constant, I'll change it.

> I'm not sure this message adds much, and if we also add a call to upgrade_charm in reconcile ([here](https://github.com/canonical/istio-operators/pull/365/files#diff-d3a4e684bd1a3a1822df5ab7605cfce00c97cd7b52f44ed770526c4f4f80eacdR293)), the logs are going to have this warning on every config-changed, relation-joined, relation-changed, etc. That feels like a lot

I want to be very verbose about this, actually, I think it must be clear what kind of installation users are receiving and why.

Reply:
I think this conversation and [the above thread](https://github.com/canonical/istio-operators/pull/365#discussion_r1443356485) have now merged - I'll keep replying to the other and we can apply whatever we decide there here too